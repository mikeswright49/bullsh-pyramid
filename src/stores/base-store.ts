import firebase from 'firebase';
import { getFirebaseConfig } from 'config/firebase-config';

export abstract class BaseStore {
    public static database: firebase.database.Database;

    public static init() {
        if (!firebase.apps.length) {
            firebase.initializeApp(getFirebaseConfig());
            firebase.auth().signInAnonymously();
        }
        if (!BaseStore.database) {
            BaseStore.database = firebase.database();
        }
    }
}
