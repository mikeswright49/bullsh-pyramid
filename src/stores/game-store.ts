import * as firebase from 'firebase';
import { shortId } from '../utilities/shortid';
import { GameStage } from '../enums/game-stage';
import { GameState } from 'types/game-state';
import { getFirebaseConfig } from '../../config/firebase-config';
import { Player } from 'types/player';
import { Card } from 'types/card';

export class GameStore {
    private static database: firebase.database.Database;
    private static subscribers = [];

    public static async createGame(playerCount: number, tierCount: number): Promise<string> {
        GameStore.init();
        const ID_LENGTH = 4;
        const gameId = shortId(ID_LENGTH);

        const gameState = {
            id: gameId,
            gameStage: GameStage.Initiation,
            activeIndex: 0,
            activeRow: 0,
            playerCount,
            tierCount,
        } as GameState;

        try {
            await GameStore.database.ref(`games/${gameId}`).set(gameState);
            return gameId;
        } catch (error) {
            console.error(error);
        }
    }

    public static subscribeToGame(gameId: string, updateCallback: (snapshot: GameState) => void) {
        GameStore.init();
        const game = GameStore.database.ref(`games/${gameId}`);
        GameStore.subscribers.push(
            game.on('value', (val) => {
                const value = val.val();
                updateCallback(value as GameState);
            })
        );
    }

    public static subscribeToPlayers(gameId: string, updateCallback: (snapshot: Player) => void) {
        GameStore.init();
        const game = GameStore.database.ref(`games/${gameId}/players`);
        const players = GameStore.database.ref(`players`);
        game.on('child_added', (child) => {
            const playerId = child.val();
            players.child(playerId).on('value', (val) => {
                const value = val.val();
                updateCallback(value as Player);
            });
        });
    }

    public static unsubscribeToGame() {
        GameStore.subscribers.forEach((sub) => sub.off('value'));
    }

    public static async updateGameStage(gameId: string, gameStage: GameStage) {
        await GameStore.database.ref(`games/${gameId}`).update({
            gameStage,
        });
    }

    public static async updateTiers(gameId: string, tiers: Card[][]) {
        await GameStore.database.ref(`games/${gameId}`).update({
            tiers,
        });
    }

    public static async setActiveCard(
        gameId: string,
        index: { activeRow: number; activeIndex: number }
    ) {
        await GameStore.database.ref(`games/${gameId}`).update(index);
    }

    public static async joinGame(gameId: string, playerId: string) {
        await GameStore.database.ref(`games/${gameId}/players`).push(playerId);
    }

    private static init() {
        if (!firebase.apps.length) {
            firebase.initializeApp(getFirebaseConfig());
        }
        if (!GameStore.database) {
            GameStore.database = firebase.database();
        }
    }
}
