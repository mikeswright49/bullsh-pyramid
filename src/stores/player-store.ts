import * as firebase from 'firebase';
import { getFirebaseConfig } from '../../config/firebase-config';
import { shortId } from '../utilities/shortid';
import { Card } from 'types/card';
import { Player } from 'types/player';

export class PlayerStore {
    private static DEFAULT_PLAYER_SCORE = 10000;
    private static database: firebase.database.Database;
    private static subscribers = [];

    public static subscribeToPlayer(playerId: string, updateCallback: (snapshot: Player) => void) {
        const player = PlayerStore.database.ref(`players/${playerId}`);
        PlayerStore.subscribers.push(
            player.on('value', (val: { val: () => Player }) => {
                const value = val.val();
                updateCallback(value as Player);
            })
        );
    }

    public static subscribeToHaters(playerId: string, updateCallback: (hater: Player) => void) {
        const player = PlayerStore.database.ref(`players/${playerId}/hatersmap`);
        const players = PlayerStore.database.ref(`players`);
        PlayerStore.subscribers.push(
            player.on('child_added', (child) => {
                const playerId = child.val();
                players.child(playerId).once('value', (val) => {
                    const value = val.val();
                    updateCallback(value as Player);
                });
            })
        );
    }

    public static async updatePlayer(player: Player) {
        await PlayerStore.database.ref(`players/${player.id}`).update(player);
    }

    public static unsubscribeToPlayer() {
        PlayerStore.subscribers.forEach((sub) => sub.off('value'));
    }

    public static async createPlayer(name: string): Promise<string> {
        const playerId = shortId();

        const player: Player = {
            id: playerId,
            name,
            hand: [],
            score: PlayerStore.DEFAULT_PLAYER_SCORE,
            isHost: false,
            declaration: null,
            hasVoted: false,
            haters: [],
            hatersmap: {},
        };

        try {
            await PlayerStore.database.ref(`players/${playerId}`).set(player);
            return playerId;
        } catch (error) {
            console.error(error);
        }
    }

    public static async setDeclaration(playerId: string, card: Card) {
        try {
            await PlayerStore.database.ref(`players/${playerId}`).update({
                declaration: card,
                hasVoted: true,
            });
        } catch (e) {
            console.error(e);
        }
    }

    public static async setHasVoted(playerId: string, hasVoted: boolean) {
        try {
            await PlayerStore.database.ref(`players/${playerId}`).update({
                hasVoted,
            });
        } catch (e) {
            console.error(e);
        }
    }

    public static async addHater(playerId: string, haterId: string) {
        try {
            await PlayerStore.database.ref(`players/${playerId}/hatersmap`).push(haterId);
        } catch (e) {
            console.error(e);
        }
    }

    public static init() {
        if (!firebase.apps.length) {
            firebase.initializeApp(getFirebaseConfig());
        }
        PlayerStore.database = firebase.database();
    }
}
