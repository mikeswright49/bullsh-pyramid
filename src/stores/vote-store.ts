import { BaseStore } from './base-store';
import { shortId } from 'src/utilities/shortid';
import { Vote } from 'types/vote';
import { Card } from 'types/card';
import { PlayerStore } from './player-store';
import { Player } from 'types/player';

export class VoteStore extends BaseStore {
    public static async createVote({
        amount,
        playerId,
        targetId,
        card,
    }: {
        amount: number;
        playerId: string;
        targetId: string;
        card: Card;
    }): Promise<string> {
        const voteId = shortId();
        try {
            await VoteStore.database.ref(`votes/${voteId}`).set({ amount, card });
            await VoteStore.database.ref(`votes/${voteId}/playerref`).push(playerId);
            await VoteStore.database.ref(`votes/${voteId}/targetref`).push(targetId);
            return voteId;
        } catch (error) {
            console.error(error);
        }
    }

    public static async addVote(gameId: string, voteId: string) {
        try {
            await VoteStore.database.ref(`/games/${gameId}/votes`).push(voteId);
        } catch (error) {
            console.error(error);
        }
    }

    public static async getTarget(voteId: string): Promise<Player> {
        const ref = VoteStore.database.ref(`votes/${voteId}/targetref`);
        return VoteStore.getPlayerRef(ref);
    }

    public static async getPlayer(voteId: string): Promise<Player> {
        const ref = VoteStore.database.ref(`votes/${voteId}/playerref`);
        return VoteStore.getPlayerRef(ref);
    }

    public static async removeVotes(gameId: string) {
        try {
            const votes = await VoteStore.database.ref(`/games/${gameId}/votes`).once('value');
            await Promise.all(
                votes.val().map((childId: string) => {
                    return votes.child(childId).ref.remove();
                })
            );
        } catch (error) {
            console.error(error);
        }
    }

    public static subscribeToVotes(gameId: string, callback: (vote: Vote) => void) {
        if (!gameId) {
            return;
        }
        const game = VoteStore.database.ref(`games/${gameId}/votes`);
        const votes = VoteStore.database.ref(`votes`);
        game.on('child_added', (child) => {
            const voteId = child.val();
            votes.child(voteId).on('value', async (val) => {
                const value: Vote = val.val();
                value.id = voteId;
                value.player = await VoteStore.getPlayer(voteId);
                value.target = await VoteStore.getTarget(voteId);
                callback(value as Vote);
            });
        });
    }

    private static async getPlayerRef(ref: firebase.database.Reference): Promise<Player> {
        const playerId: string = (await ref.once('child_added')).val();
        return PlayerStore.getPlayer(playerId);
    }
}
