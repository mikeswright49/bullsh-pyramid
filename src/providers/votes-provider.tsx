import React from 'react';
import { useState, useEffect } from 'react';
import { Vote } from 'types/vote';
import { VoteStore } from 'src/stores/vote-store';
export const VotesContext = React.createContext<Vote[]>([]);

export function VotesProvider(props: React.PropsWithChildren<{ gameId: string }>) {
    const [votes, setVotes] = useState([]);
    useEffect(() => {
        if (!votes) {
            return;
        }

        VoteStore.subscribeToVotes(props.gameId, (vote: Vote) => {
            votes[vote.id] = vote;
            setVotes({ ...votes });
        });
    }, [props.gameId]);

    return (
        <VotesContext.Provider value={Object.values(votes)}>{props.children}</VotesContext.Provider>
    );
}
