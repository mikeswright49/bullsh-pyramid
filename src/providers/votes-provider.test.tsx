import React from 'react';
import { useContext } from 'react';
import { render, act } from '@testing-library/react';
import { VoteStore } from 'src/stores/vote-store';
import { VotesContext, VotesProvider } from './votes-provider';
import { MOCK_VOTE } from 'src/__tests__/data/mock-vote';

jest.mock('src/stores/vote-store');

describe('<Unit Test> Votes Provider', () => {
    function ComponentMock() {
        const data = useContext(VotesContext);
        return <div>{JSON.stringify(data)}</div>;
    }

    it('should subscribe to the votes from the store, and map the votes to an array it as a value', async () => {
        let cb;
        (VoteStore.subscribeToVotes as jest.Mock) = jest.fn((id, _cb) => (cb = _cb));

        const { container } = render(
            <VotesProvider gameId="123">
                <ComponentMock />
            </VotesProvider>
        );

        act(() => cb(MOCK_VOTE));

        expect(VoteStore.subscribeToVotes).toHaveBeenCalledWith('123', expect.any(Function));
        expect(container).toMatchSnapshot();
    });
});
