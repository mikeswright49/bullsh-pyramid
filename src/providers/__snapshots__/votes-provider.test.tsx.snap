// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Unit Test> Votes Provider should subscribe to the votes from the store, and map the votes to an array it as a value 1`] = `
<div>
  <div>
    [{"id":"abc","accepted":false,"bullshit":false,"turnKey":"1-2","amount":1,"card":{"hidden":true,"suit":1,"value":12},"playerRef":{"-MPRrYkfedrmx5dF3KPD":"ucx"},"targetRef":{"-MPRrYlH53OR8GC58EJO":"opb"},"player":{"hand":[{"hidden":true,"suit":2,"value":7},{"hidden":true,"suit":3,"value":8},{"hidden":true,"suit":1,"value":11},{"hidden":true,"suit":2,"value":0}],"id":"nez","isHost":true,"name":"frank","score":10000,"declaration":[],"voteSubmitted":0},"target":{"hand":[{"hidden":true,"suit":0,"value":0},{"hidden":true,"suit":2,"value":2},{"hidden":true,"suit":1,"value":6},{"hidden":true,"suit":2,"value":10}],"id":"opb","isHost":false,"name":"frank","score":10000}}]
  </div>
</div>
`;
